oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp

```diff
@@ -0,0 +1,276 @@
+#include "pfw_sms_menu.hpp"
+
+#include <security/pam_appl.h>
+
+#include <cstdlib>
+#include <cstring>
+
+namespace ibm_pfw_sms
+{
+
+/** @struct Data for the PAM Conversation function
+ *
+ * @note This is passed from our client code through PAM into the conversation
+ * via appdata_ptr
+ *
+ * [in] password - this the password that our application
+ * already has before invoking PAM. For authentication, this is the current
+ * password. For password change, this is the proposed new password.
+ * [out] changePasswordReasonCode - If one of the PAM modules (such as
+ * pam_pwquality) gives a response, we convert it to a reason code and store it
+ * here. The caller will need examine both this code and the return value from
+ * the pam_* function.
+ */
+struct pfw_sms_pam_appdata
+{
+    const std::string& password;
+    enum changePasswordReasonCode reasonCode;
+};
+
+/** @brief Our PAM Conversation function
+ *
+ * This function handles the following elements of the PAM conversation:
+ *  - For both authentication and password change, when PAM asks for the
+ * password, we supply it.
+ *  - For the change password function, if a module (such as pam_pwquality.so)
+ * complains about the password quality, we translate that message into a reason
+ * code (appdata_ptr->reasonCode). We use struct pfw_sms_pam_appdata (as this
+ * function's appdata_ptr) to pass data to and from our application and PAM:
+ * @note For parameter descriptions, see docs for pam_start and pam_conv
+ * @return PAM status code (int)
+ */
+static int pamConversationFunction(int num_msg,
+                                   const struct pam_message** message_ptr,
+                                   struct pam_response** response_ptr,
+                                   void* appdata_ptr)
+{
+    if ((appdata_ptr == NULL) || (message_ptr == NULL) ||
+        (response_ptr == NULL))
+    {
+        return PAM_CONV_ERR;
+    }
+    if (num_msg <= 0 || num_msg >= PAM_MAX_NUM_MSG)
+    {
+        return PAM_CONV_ERR;
+    }
+
+    struct pfw_sms_pam_appdata* pfw_sms_appdata_ptr =
+        (struct pfw_sms_pam_appdata*)appdata_ptr;
+    *response_ptr = (pam_response*)calloc(
+        num_msg, sizeof(struct pam_response)); // PAM frees
+    if (*response_ptr == NULL)
+    {
+        return PAM_CONV_ERR;
+    }
+    const struct pam_message* msg_ptr = *message_ptr;
+    for (int msg_index = 0; msg_index < num_msg; ++msg_index)
+    {
+        char* local_password = NULL; // for PAM's malloc'd copy
+        std::basic_string_view msg(msg_ptr[msg_index].msg);
+        switch (msg_ptr[msg_index].msg_style)
+        {
+            case PAM_PROMPT_ECHO_OFF:
+                // Assume PAM is asking for the password.  Supply a malloc'd
+                // password that PAM will free
+                local_password = ::strdup(
```

Comment:
please use the std::string class , no c-string manipulation API's.

Reply:
Sorry, I cannot do that.  The Linux-PAM code requires my code to malloc() (or calloc()) and it performs free(), anbd I cannot change Linux-PAM, so we are locked into this old-style memory management.  I added a comment in several places to explain why we must do this.

Reply:
ah okay , then having comments to make people understand reason is good enough. Thanks joseph.